ext/drpttracer/drpttracer.c

```diff
@@ -0,0 +1,579 @@
+        dr_global_free(buf, buf_size);
+        return false;
+    }
+    *config |= (val & mask) << start;
+
+    dr_global_free(buf, buf_size);
+    return true;
+}
+
+/***************************************************************************
+ * PERF EVENT RELATED FUNCTIONS
+ */
+
+/* This function is used to initialize one perf_event_attr.
+ * For kernel only tracing:
+ *  The perf_event_attr is set to be the same as running
+ *  'perf record -e intel_pt/noretcomp/k'
+ * For user only tracing:
+ *  The perf_event_attr is set to be the same as running
+ *  'perf record -e intel_pt/cyc,noretcomp/u'
+ * For both kernel and user tracing:
+ *  The perf_event_attr is set to be the same as running
+ *  'perf record -e intel_pt/cyc,noretcomp/uk'
+ */
+static bool
+pt_perf_event_attr_init(bool user, bool kernel, OUT struct perf_event_attr *attr)
+{
+    memset(attr, 0, sizeof(struct perf_event_attr));
+    attr->size = sizeof(struct perf_event_attr);
+    if (!parse_pt_pmu_type(&attr->type)) {
+        ERRMSG("failed to parse PT's pmu type\n");
+        return false;
+    }
+    uint64_t config = 0;
+    attr->exclude_user = 1;
+    attr->exclude_hv = 1;
+    if (!parse_pt_pmu_event_config("noretcomp", 1, &config)) {
+        ERRMSG("failed to parse PT's pmu event noretcomp to perf_event_attr.config\n");
+        return false;
+    }
+    if (user) {
+        attr->exclude_kernel = 1;
+        attr->exclude_hv = 1;
+        if (!parse_pt_pmu_event_config("cyc", 1, &config)) {
+            ERRMSG("failed to parse PT's pmu event cyc to perf_event_attr.config\n");
+            return false;
+        }
+    }
+    attr->config = config;
+    attr->disabled = 1;
+    return true;
+}
+
+/* This function is used to open a perf event.
+ */
+static int
+perf_event_open(IN struct perf_event_attr *attr, IN pid_t pid, IN int cpu,
+                IN int group_fd, IN unsigned long flags)
+{
+    errno = 0;
+    int fd = syscall(SYS_perf_event_open, attr, pid, cpu, group_fd, flags);
+    if (fd < 0 || errno != 0) {
+        ERRMSG("failed to open perf event: %s\n", strerror(errno));
+        return -1;
+    }
+    return fd;
+}
+
+/***************************************************************************
+ * PT META DATA RELATED FUNCTIONS
+ */
+
+/* This function is used to get the PT's shared meta data:
+ *  cpu_family, cpu_model, cpu_step
+ */
+static void
+pt_shared_metadata_init(pt_metadata_t *metadata)
+{
+    memset(metadata, 0, sizeof(pt_metadata_t));
+    int cpu_family = 0, cpu_model = 0, cpu_stepping = 0;
+    get_cpu_info(&cpu_family, &cpu_model, &cpu_stepping);
+    metadata->cpu_family = (uint16_t)cpu_family;
+    metadata->cpu_model = (uint8_t)cpu_model;
+    metadata->cpu_stepping = (uint8_t)cpu_stepping;
+}
+
+/***************************************************************************
+ * PT TRACER HANDLE RELATED FUNCTIONS
+ */
+
+/* The type of PT trace's handle. The handle is used to store information about the
```

Comment:
s/The type of PT trace's handle./The PT trace handle/

Reply:
Done.